sql/rpl_binlog_sender.cc

```diff
@@ -528,6 +537,11 @@ int Binlog_sender::send_events(File_reader *reader, my_off_t end_pos) {
     processlist_slave_offset(log_file, log_pos);
 
     if (before_send_hook(log_file, log_pos)) DBUG_RETURN(1);
+
+    if (rpl_wait_for_semi_sync_ack && m_is_semi_sync_slave &&
+        is_event_executed(event_ptr))
+      mysql_bin_log.signal_semi_sync_ack(log_file, log_pos);
```

Comment:
AFAIK (at least in 5.6) gtid_executed can contain un-acked transactions because it's updated after the flush stage in group commit. So this might not be correct.

Reply:
We're signalling on the GTID event, so the rest of the transaction will not be signaled, correct?

Reply:
This code calls `signal_semi_sync_ack()` only for the following events:
```
SET @@SESSION.GTID_NEXT=`...`
```
It means that in case of server restart and a semi-sync slave starts asking for binlogs, async slaves will be updated only to the start of the last transaction. The rest of the transaction will be updated when new writes come into the system. 
Another issue is that when new writes come into the system async slaves can be updated up to the start of the last transaction which may be not acked yet.

Reply:
The first issue can be fixed, but the second (updating up to the start of the last transaction which may be not acked yet) is hard to fix.